Implement SCIM filtering

S203932 opened the pull request:
Allowing for filtering according to the following Operators: 
- eq
- co
And with the following attributes: 
- familyName
- givenName
- username
- active
- emails

rmiccoli committed:
Grant admin scopes to admin-approved clients only

enricovianello committed:
Bump version to 1.11.0

darcato committed:
Self-service upload of user certificate (#750)

- Both the user and the admin can link a certificate to their account
- The admin must approve certificate linking requests from other users
- The certificate can be added via PEM or subject and issuer
- The issuer can be chosen among a list of "known" CAs (usually IGTF and system certificates).
- Both subject and issuer DN format will be checked to be compliant with the comma-separated standard (RFC 2253)

---------
Co-authored-by: rmiccoli <roberta.miccoli@cnaf.infn.it>

garaimanoj committed:
Update matchingPolicy

rmiccoli committed:
Handle HEAD requests

for /registration/verify/{token} endpoint.

Return 405 status code

Test HEAD request

federicaagostini committed:
Redirect to login page when signing AUP

rmiccoli committed:
Bump version to 1.10.3

rmiccoli committed:
Bump mitreid version to 1.3.7.cnaf-20241119

rmiccoli committed:
Fix tests

SteDev2 committed:
Add confirmation before rotate client secret

rmiccoli committed:
Fix account mapping

* Fix account mapping when linking x509 certificates
* Fix account mapping in VOMS-AA before generating an AC

---------

Co-authored-by: Enrico Vianello <enrico.vianello@cnaf.infn.it>

enricovianello committed:
Add POST endpoint for registration requests confirmation (#881)

* Add POST endpoint for registration requests confirmation
* Remove stack trace from log
* Remove unused confirm endpoint

enricovianello committed:
Fix CERN lifecycle handler (#871)

* Consider null end date as unlimited membership
* Manage 404 response from hr db api service
* Configurable behavior in case VO Person or CERN person id are not found
* Simplify API exceptions management

enricovianello committed:
Revert "Self-service upload of user certificate (#750)"

This reverts commit ae09f3938d82976f0a9e400af8fdbdcebc4e0a9e.

enricovianello committed:
Revert "Bump version to 1.11.0"

This reverts commit 22faeb3e86fcf45ca418b09b1a98874827769d32.

enricovianello committed:
Cosmetic fixes before 1.10.3 release (#882)

* Log level reduced from info to debug
* Check HEAD on registration confirm endpoint returns 200

enricovianello committed:
Bump version to 1.11.0

enricovianello committed:
Prevent the issue of broken SAML login flow

rmiccoli committed:
Exclude IAM optional groups from VOMS AC (#894)

DonaldChung-HK committed:
Fix #809 JDBC session on same db (#898)

enricovianello committed:
Fix CERN lifecycle logic (#896)

- set end time to current date in case user is not found on API
- set ent time to current date when user has no experiment participation at all

rmiccoli committed:
Find account by certificate sub and iss in VOMS AA (#897)

enricovianello committed:
Avoid NullPointerException

rmiccoli committed:
Fix code smells

enricovianello committed:
Remove Institute label

The 64 char limit on label val is not enough to store Institute.

enricovianello committed:
Log lifecycle handlers start and end with INFO level

garaimanoj committed:
Delete rogue registration requests without sending a reason (#858)

* Add motivation to the message published into RegistrationRejectEvent

rmiccoli committed:
Add MFA support for local credentials (#733)

enricovianello committed:
Add 1.11.0 to CHANGELOG

enricovianello committed:
Remove transient keywords to avoid missing info stored in session

enricovianello committed:
Fix lifecycle management

enricovianello committed:
Fix AuthZ on endpoints that requires identity

enricovianello committed:
Fix infinite recursion on object serialization

enricovianello committed:
Fix sonar issues

enricovianello committed:
Fix sonar issues

enricovianello committed:
Add validation to Attribute DTO and fix Sonar issues

enricovianello committed:
Fix errors after rebase

enricovianello committed:
Latest CERN lifecycle fixes

federicaagostini committed:
Create an approved site when using device code flow (#908)

* Proper error message for not-allowed scopes
* Fix AuthorizationPendingException error when client was already approved

enricovianello committed:
Combine all scope filtering logic into one ScopeFilter (#912)


---------

Co-authored-by: Federica Agostini <federica.agostini@cnaf.infn.it>

enricovianello committed:
Remove conflicting bean definition and improve exception debugging on main

rmiccoli committed:
Fix sonar issues

federicaagostini committed:
Add client details before consent page

rmiccoli committed:
Add message on the user code page

enricovianello committed:
Fix Authz on controllers and cosmetic fixes

enricovianello committed:
Fix for https://github.com/indigo-iam/iam/issues/866

enricovianello committed:
Fix sonar issues and cosmetic fixes

enricovianello committed:
Restrict access to SCIM scopes to admins only

Also client credentials flow is still allowed

enricovianello committed:
Add SecurityExpressions tests

enricovianello committed:
Bump version to 1.12.0

Commit:
Simple filtering added. Only prove of concept. Imidiate revision is required. Still under development.

Commit:
Simple filtering added. Only prove of concept. Imidiate revision is required. Still under development.

Commit:
Simple filtering added. Only prove of concept. Imidiate revision is required. Still under development.

Commit:
Functioning implementation of filtering with attributes. Minor bugs still existing.

Commit:
Functioning implementation of filtering with attributes. Minor bugs still existing.

Commit:
Functioning implementation of filtering with attributes. Minor bugs still existing.

Commit:
Functional filtering with valid counter to show total search results. Minor bug when combining it with count as it currently limits the search

Commit:
Functional filtering with valid counter to show total search results. Minor bug when combining it with count as it currently limits the search

Commit:
Functional filtering with valid counter to show total search results. Minor bug when combining it with count as it currently limits the search

Commit:
Funtional filtering with pagination. Pagnination with filtering refers to the results from the filtering and not all the users.

Commit:
Funtional filtering with pagination. Pagnination with filtering refers to the results from the filtering and not all the users.

Commit:
Funtional filtering with pagination. Pagnination with filtering refers to the results from the filtering and not all the users.

Commit:
Removed some commented-out code

Commit:
Removed some commented-out code

Commit:
Fixed search error. Invalid search parameters causes know gives an empty response

Commit:
removed out-commented code and unused libraries

Commit:
removed out-commented code and unused libraries

Commit:
Change to query data based on filter, rather than gathering all data and then filter. Only one query at the moment and count and startIndex is not implemented

Commit:
Change to query data based on filter, rather than gathering all data and then filter. Only one query at the moment and count and startIndex is not implemented

Commit:
Change to query data based on filter, rather than gathering all data and then filter. Only one query at the moment and count and startIndex is not implemented

Commit:
Removed customList from ScimProvisioning interface as the method is currently not to be implemented in ScimGroupProvisioning

Commit:
Removed customList from ScimProvisioning interface as the method is currently not to be implemented in ScimGroupProvisioning

Commit:
Removed customList from ScimProvisioning interface as the method is currently not to be implemented in ScimGroupProvisioning

Commit:
Changed filtering of SCIM api to only fetch the correct data based on the filter rather than fetching all data and then filter

Commit:
Changed filtering of SCIM api to only fetch the correct data based on the filter rather than fetching all data and then filter

Commit:
Changed filtering of SCIM api to only fetch the correct data based on the filter rather than fetching all data and then filter

Commit:
Fixed minor bug, contains shouldn't be allowed with the active attribute as it is a boolean

Commit:
Fixed minor bug, contains shouldn't be allowed with the active attribute as it is a boolean

Commit:
Fixed minor bug, contains shouldn't be allowed with the active attribute as it is a boolean

garaimanoj committed:
Add a READER role that gives read access to users and groups info (#842)

* Add a new authority ROLE_READER
* Add Assign monitoring privileges button on the user homepage
* Hide group request and ssh key from reader
* Test API access

garaimanoj committed:
AUP exemption for service accounts (#811)

federicaagostini committed:
Assign oidc-agent clients ownership to the whom approved it (#917)

garaimanoj committed:
Integrate MFA with x.509 authentication (#926)

federicaagostini committed:
Allow to customize the email subject prefix (#941)

federicaagostini committed:
Add endpoint to search users by their authority (#940)

Commit:
Added test for filtering in ScimUserProvisioning

Commit:
Added test for filtering in ScimUserProvisioning

S203932 committed:
Merge branch 'develop' into issue-931

Commit:
added header to the filtering test file

Commit:
added header to the filtering test file

Commit:
I had made an error in previous merge in ScimUserController, should be fixed now

Commit:
I had made an error in previous merge in ScimUserController, should be fixed now

Commit:
I had made an error in previous merge in ScimUserController, should be fixed now

Commit:
Removed unnecessary imports in tests

Commit:
Removed unnecessary imports in tests

Commit:
now formatted

Commit:
now formatted

Commit:
now formatted

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -112,59 +112,56 @@ public ScimUserProvisioning(IamAccountService accountService,
           int index = filtersParameter.indexOf(operator.toString(), from);
 
-          index = index==0?index=1:index;
+          index = index == 0 ? index = 1 : index;
```
this I don't understand. does it mean that if index is 0, it should become 1? maybe an `if` is clearer. and in any case I don't think you need `index = ` before `1`.

S203932 replied:
Yes, if index is equal to 0 than it should become 1. It is to make sure it does not get stuck by finding the next instance and keep moving it forward. 
Yes, one can omit the `index = `  before the `1`, haven't been using shorthand that much :sweat_smile:
I can do an `if` if that is preferred, just did the shorthand as it is shorter, line-wise

giacomini replied:
In this case I think an `if` would be clearer
```suggestion
          if (index == 0) {
            index = 1;
          };
```

S203932 replied:
Sure thing, I commited your suggestion

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -174,12 +171,12 @@ private ArrayList<String> parseFilters(final String filtersParameter) {
-          for(AttributeOperators operator : AttributeOperators.values()){
-            if(operator.toString().equalsIgnoreCase(parsedFilters.get(1))){
+  private boolean filterEvaluation(ArrayList<String> parsedFilters) {
+    if (parsedFilters.size() == 3) {
```
can we reduce the overall nesting? maybe with an early return if parsedFilters.size() is not 3?

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -112,59 +112,56 @@ public ScimUserProvisioning(IamAccountService accountService,
+            String value = filtersParameter.substring(index + operator.name().length() + 1,
+                filtersParameter.length());
+
             result.add(parameter);
```
This code is probably pre-existing, but I wonder if it would be better to represent a filter not as an array of three strings (which implies using get(i)), but as as a struct/class with three named members. Sorry for using a C++ terminology.

giacomini replied:
I leave this to @enricovianello to decide if he wants to profit from this opportunity to restructure the code and how.

S203932 replied:
> Personally, I prefer this approach as it keeps the code cleaner and more readable. You could create the Filter class (e.g., ScimFilter.class) under /iam/api/scim/model.

The Arraylist has now been replaced with the ScimFilter object from the class in  `/iam/api/scim/model/ScimFilter.java

rmiccoli committed:
Show the number of active users (#943)

enricovianello committed:
Fix missing delete label event (#950)

S203932 committed:
Update iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java

Co-authored-by: Francesco Giacomini <francesco.giacomini@cnaf.infn.it>

S203932 committed:
Update iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java

Co-authored-by: Francesco Giacomini <francesco.giacomini@cnaf.infn.it>

S203932 committed:
Update iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java

Co-authored-by: Francesco Giacomini <francesco.giacomini@cnaf.infn.it>

S203932 committed:
Update iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java

Co-authored-by: Francesco Giacomini <francesco.giacomini@cnaf.infn.it>

S203932 committed:
Minor change to pull request correction in method filterEvalutation from file ScimUserProvisioning.java

S203932 committed:
Minor change to pull request correction in method filterEvalutation from file ScimUserProvisioning.java

S203932 committed:
Found and fixed a bug in parseFilters from ScimUserProvisioning. Possible infinit loop bug

S203932 committed:
Found and fixed a bug in parseFilters from ScimUserProvisioning. Possible infinit loop bug

S203932 committed:
Changed the filter in ScimUserProvisioning from using an arraylist to use a Filter.class

S203932 committed:
Changed the filter in ScimUserProvisioning from using an arraylist to use a Filter.class

enricovianello committed:
Use setup-java@v4

federicaagostini committed:
Add support for resource parameter (#916)

- Add the requested resources to the consent page
- Allow iam-test-client to perform 'resource' request
- Add all granted resources if not requested at token endpoint 
- All granted resources can be requested also with aud or audience parameters at token endpoint

S203932 committed:
Changed some names and removed some redundancy to redeuce the number of Sonar issues

S203932 commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -112,59 +112,56 @@ public ScimUserProvisioning(IamAccountService accountService,
 
           // If there is an occurence and it's surrounded by spaces
-          if(index > 0 && filtersParameter.charAt(index-1) == ' ' && filtersParameter.charAt(index + operator.name().length()) == ' '){
+          if (index > 0 && filtersParameter.charAt(index - 1) == ' '
```
@giacomini I have no removed the check. 

Yes `index` could be `-1` if there was no instance of the given `operator` in the filtering string. 
However, this is not possible given that the code does not enter the `while` - loop in line `140` if that is the case, i.e. it should not be possible for `index` to be `-1` and the given `if` statement.

S203932 committed:
Add support for voperson_id claim in the AARC JWT profile (#958)

balciiberk committed:
Support DER & URL-Encoded PEM client certificates (#938)

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/model/ScimFilter.java:
```diff
@@ -0,0 +1,47 @@
+
+public class ScimFilter {
+
+    private final String attribute;
```
We could use enum as types for both `attribute` and `operator`. The enum defined inside ScimUserProvisioning class

S203932 replied:
It has been done :+1: 

`ScimFilter` now uses the enums `ScimFilterAttributes` and `ScimFilterOperators`

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -107,6 +111,163 @@ public ScimUserProvisioning(IamAccountService accountService,
+    // Other info
+    active, emails
+  }
+
```
I always prefer having types in separated files. I'd move these enums outside this class. If missing, we can create a sub-package inside it.infn.mw.iam.api.scim.provisioning called model for example.

S203932 replied:
It has been done :+1: 
They have been moved to their own enum classes in the folder iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/model/`. 
The classes are `ScimFilterAttributes` and `ScimFilterOperators`

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -107,6 +111,163 @@ public ScimUserProvisioning(IamAccountService accountService,
+
+
+  private ScimFilter parseFilters(final String filtersParameter) {
+
```
My suggestion here is to create a ScimFilterParser that implements a method like:

```
public ScimFilter parse(String filter)
```

We can do this parsing with less if-then-else and substring calls in order to obtain a more readable codebase.
We can use regex to parse the expected string, for example:

```
  String regex = "([\\w\\.]+)\\s+(eq|co)\\s+\"([^\"]*)\"";
  Pattern pattern = Pattern.compile(regex);
  Matcher matcher = pattern.matcher(filtersParameter);

  if (!matcher.matches()) {
    throw invalidFilter(filtersParameter);
 }
```

Then you can use the matcher to get the String parts of the ScimFilter:

```
  String attributeStr = matcher.group(1);
  String operatorStr = matcher.group(2);
  String value = matcher.group(3);
```
and use the Enum native methods to get the correct Attribute/Operator and create the ScimFilter object:

```
  Attribute attribute = parseAttribute(attributeStr);
  AttributeOperator operator = AttributeOperator.valueOf(operatorStr);
  return new ScimFilter(attribute, operator, value);
```

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -107,6 +111,163 @@ public ScimUserProvisioning(IamAccountService accountService,
+  }
+
+
+  private boolean filterEvaluation(ScimFilter parsedFilters) {
```
this method becomes useless if we use enums and is no more necessary to evaluate the ScimFilter object

S203932 replied:
It has been removed :+1:

enricovianello reviewed (changes requested):
I'm suggesting things that should make codebase more readable. It's the start of a review but I think I've suggested a good amount of things and then I'll have a second review as soon as these changes are in place. Sorry for this amount of changes :)

rmiccoli committed:
Allow users to see group's labels from dashboard (#944)

enricovianello committed:
Support no mandatory notes on registration (#948)

---------
Co-authored-by: Sae126V <saitejav2021@gmail.com>
Co-authored-by: rmiccoli <roberta.miccoli@cnaf.infn.it>

enricovianello committed:
Change the message when identifying with X509 and the account is suspended (#965)

---------
Co-authored-by: rmiccoli <roberta.miccoli@cnaf.infn.it>

garaimanoj committed:
Avoid persisting too long request parameters (#957)

mmutadich committed:
Add PATCH replace operation to SCIM Group endpoint (#955)

enricovianello committed:
Upgrade dependencies and minor cleanups (#966)

* Upgrade VOMS API Java dependency to latest 3.3.5
* Upgrade nimbus-jose-jwt version to 9.37.2
* Remove duplicate declaration
* Simplify MFA verify model

enricovianello committed:
Log the outcome of a VOMS proxy request (#921)

S203932 committed:
Merge branch 'issue-931' of https://github.com/indigo-iam/iam into issue-931

S203932 committed:
Revert "Use setup-java@v4"

This reverts commit 748724f97b64839e7d4698a4c8ac05cee63fcbd4.

S203932 committed:
Reapply "Use setup-java@v4"

This reverts commit 6f985584f6c185002116ef8cb7e61e52d3ab191c.

S203932 committed:
Fixed error from bad rebase/merge

S203932 committed:
Removed duplicate if statement in scimusercontroller

S203932 committed:
Removed duplicate if statement in scimusercontroller

S203932 committed:
Fixed the project again after a squash attempt that ended up putting 11 additional commits instead

S203932 committed:
Revert "Fixed the project again after a squash attempt that ended up putting 11 additional commits instead"

This reverts commit 8a7c8c7ceae29a24be4d9d1f98c1c51abb33ecdc.

S203932 committed:
Revert "Removed duplicate if statement in scimusercontroller"

This reverts commit 61e6ead27f22a246c7c4523aa0d46b150be7fea5.

S203932 committed:
Revert "now formatted"

This reverts commit 647aaef8a37ffb6b6669cb03f50b3b630542eaf8.

S203932 committed:
Revert "I had made an error in previous merge in ScimUserController, should be fixed now"

This reverts commit 7a12c349d7756d8a911bb9d729c768a7c7f0baf3.

S203932 committed:
Revert "Fixed minor bug, contains shouldn't be allowed with the active attribute as it is a boolean"

This reverts commit c2538fdf5165629f53ca9c181fe3b7e1aed8a312.

S203932 committed:
Revert "Changed filtering of SCIM api to only fetch the correct data based on the filter rather than fetching all data and then filter"

This reverts commit 7e0e1930a263d1ab5ca4f62fb8afe1f38845782f.

S203932 committed:
Revert "Removed customList from ScimProvisioning interface as the method is currently not to be implemented in ScimGroupProvisioning"

This reverts commit f26a52a8e0aab4275dbeaa351a49b89e9bfe51d0.

S203932 committed:
Revert "Change to query data based on filter, rather than gathering all data and then filter. Only one query at the moment and count and startIndex is not implemented"

This reverts commit 8c4cc367a461ac119a6787656cd0d991fde488dc.

S203932 committed:
Revert "Funtional filtering with pagination. Pagnination with filtering refers to the results from the filtering and not all the users."

This reverts commit 19439e7e9ce9f1ea9cb57bb9378d850770025e74.

S203932 committed:
Revert "Functional filtering with valid counter to show total search results. Minor bug when combining it with count as it currently limits the search"

This reverts commit 17d4d4fa48a7e2e3641c7495ad73c638fcf80f55.

S203932 committed:
Revert "Functioning implementation of filtering with attributes. Minor bugs still existing."

This reverts commit ccafbe842f595bec16eeb1e1000213587b4d6e6a.

S203932 committed:
Revert "Simple filtering added. Only prove of concept. Imidiate revision is required. Still under development."

This reverts commit 246a4114c833baa8550fe02466a4ce6f304d22f9.

S203932 committed:
changed tests as they did not take pagnation into account

S203932 committed:
Missed a pagnation value

S203932 committed:
Another pagnation error within the tests. Should be all of them now

S203932 committed:
Implemented the method customList to interface as requested. Also throwing the error and testing this scenario

S203932 committed:
In UserProvisioning changed customList to List and moved functionality of original List method to the former customList method

S203932 committed:
removed unused method ScimPageRequest in ScimControllerSupport

S203932 committed:
Fixed count=0 issue, which means that only one version of list method in ScimUserProvisioning is needed

S203932 committed:
ScimFilter attribute and operator are now its own enum classes

S203932 committed:
Parsing filter using regex

S203932 committed:
Added tests for count=0 instance

S203932 committed:
Fixing sonar cube issues

S203932 committed:
removed deprecated list method from ScimUserProvisioning

S203932 commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimProvisioning.java:
```diff
@@ -35,4 +35,7 @@ public interface ScimProvisioning<T, E> {
   ScimListResponse<T> list(ScimPageRequest params);
 
+  // The method below cannot exist until it is also implemented for ScimGroupProvisioning.java
+  // ScimListResponse<T> customList(ScimPageRequest params, String filter);
```
It has been done :+1:

S203932 commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/controller/ScimUserController.java:
```diff
@@ -79,25 +80,45 @@ private Set<String> parseAttributes(final String attributesParameter) {
+    MappingJacksonValue wrapper;
+
+
+    if (filters != null) {
```
It has been done :+1: 

The method below no longer exists.
`public ScimListResponse<ScimUser> list(pr) {
 ...
}`

The `public ScimListResponse<ScimUser> list(pr)` has been integrated into `public ScimListResponse<ScimUser> list(pr, filters)`.

`list(pr)` is still in the interface (as it is used in `ScimGroupProvisioning`), but throws an exception in `ScimUserProvisioning` (which there is a test for).

S203932 commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/controller/ScimControllerSupport.java:
```diff
@@ -28,6 +28,13 @@ protected ScimPageRequest buildUserPageRequest(Integer count, Integer startIndex
+  protected ScimPageRequest buildcustomUserPageRequest(Integer count, Integer startIndex,
+      Integer totalUsers) {
+    return buildPageRequest(count, startIndex, totalUsers);
+  }
```
Nope, it has been deleted :+1:

S203932 committed:
Removed unused import

enricovianello committed:
Merge branch 'develop' into issue-931

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimGroupProvisioning.java:
```diff
@@ -233,6 +241,11 @@ public ScimListResponse<ScimGroup> list(ScimPageRequest pageRequest) {
 
+  @Override
+  public ScimListResponse<ScimGroup> list(final ScimPageRequest params, String filter) {
+    throw methodNotSupported("list(final ScimPageRequest params, String filter)");
```
We could simplify the implementation of this by using the standard `UnsupportedOperationException`:

```
throw new UnsupportedOperationException("Unsupported filtered list");
```

Probably a simpler and better solution could be calling the `list` method defined before and ignoring the list parameter. We could add a log.warn message to track this wrong call on log.

enricovianello replied:
This makes useless and removable the ScimInvalidMethod exception

S203932 replied:
I've now changed it, so the method `list(final ScimPageRequest params, String filter)` gives a warning to the logger and then calls the `list(ScimPageRequest pageRequest)` (which doesn't have the `filter` parameter). 
The corresponding test has been changed as well to match this.

S203932 replied:
The method `list(final ScimPageRequest params, String filter)` now produces a warning in the logger and then calls the method `list(ScimPageRequest pageRequest)`, whilst ignoring the `filter` parameter. 
The corresponding test has been changed to reflect this.

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/api/scim/provisioning/ScimUserProvisioning.java:
```diff
@@ -107,6 +114,226 @@ public ScimUserProvisioning(IamAccountService accountService,
+        ScimFilterAttributes.parseAttribute(attributeStr.toLowerCase());
+    ScimFilterOperators operator = ScimFilterOperators.parseOperator(operatorStr.toLowerCase());
+
+
```
too many empty lines

S203932 replied:
They've been removed :+1:

enricovianello commented:
Conflicts to be fixed

enricovianello commented on iam-persistence/src/main/java/it/infn/mw/iam/persistence/repository/IamAccountRepository.java:
```diff
@@ -151,4 +154,62 @@ Page<IamAccount> findByLabelNameAndValue(@Param("name") String name, @Param("val
 
+
+  @Query("select a from IamAccount a join a.userInfo ui where ui.givenName = :givenName")
+  Optional<Page<IamAccount>> findByGivenName(@Param("givenName") String givenName,
```
Is really Optional necessary? Probably the returned Page when there's no results is a Page with size 0 and no contents.

S203932 replied:
You're correct, it was a bit of a barricade to make sure it would always run and not throw any errors (error handling is still in place, but a more reasonable amount). 
Optional has been removed and functions with the main use of these methods (in file `ScimUserProvisioning.java`) has been simplified.

enricovianello reviewed (changes requested):
Probably a useless Optional on repo is the main (little) change

S203932 committed:
Format + removing optional and changing to switch case

S203932 committed:
Merge branch 'develop' into issue-931

S203932 committed:
Changed list method with filter parameter and corresponding test

S203932 committed:
Removed unused imports
